### PR TITLE
literature: Add fulltext_links

### PR DESCRIPTION
**Pull request by pazembrz**
* INSPIR-2737

### CODE REVIEW

File: backend/inspirehep/records/marshmallow/base.py [hunk 1]
```diff
@@ -89,3 +89,21 @@ class MetadataSchema(EnvelopeSchema):
         metadata = fields.Nested(schema, dump_only=True)
 
     return MetadataSchema
+
+
+class RequiredFieldsSchema(Schema):
```
Comment (drjova) [inline backend/inspirehep/records/marshmallow/base.py]: IMHO this is too complex, each just a check on the specific serializer instead on ``post_dump`` or even better https://github.com/inspirehep/inspirehep/pull/498/files#diff-19d3d162ce2a08a6469fc1ae8afae85fR112

File: backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+)
+
+
+class FulltextLinkBase(RequiredFieldsSchema):
+    value = fields.Method("get_fulltext_url", dump_only=True, required=True)
+    description = fields.Method(
+        "get_fulltext_description", dump_only=True, required=True
+    )
+
+    DESCRIPTION = None
```
Comment (drjova) [inline backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py]: Maybe?
```suggestion
    default_description = None
```

File: backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            if entry["schema"].lower() == "kekscan":
+                return ExternalSystemIdentifierSchemaV1.get_link_for_kekscan_schema(
+                    value
+                )
+        return missing
+
+
+class ArxivFulltextLink(FulltextLinkBase):
+    DESCRIPTION = "arXiv"
+    ARXIV_FULLTEXT_URL = "https://arxiv.org/pdf/{arxiv_eprint_id}"
```
Comment (drjova) [inline backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py]:
```suggestion
    arxiv_fulltext_url = "https://arxiv.org/pdf/{arxiv_eprint_id}"
```

File: backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        value = entry.get("value")
+        if value:
+            return self.ARXIV_FULLTEXT_URL.format(arxiv_eprint_id=value)
+        return missing
+
+
+class InternalFulltextLink(FulltextLinkBase):
+    DESCRIPTION = "fulltext"
+
+    def get_fulltext_url(self, entry):
```
Comment (drjova) [inline backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py]:
```python
if current_app.config.get("FEATURE_FLAG_ENABLE_FILES"):
    return missing
```

File: backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        raise NotImplementedError
+
+    def get_fulltext_description(self, entry):
+        return self.DESCRIPTION
+
+
+class KekFulltextLink(FulltextLinkBase):
+    DESCRIPTION = ("KEK scanned document",)
```
Comment (drjova) [inline backend/inspirehep/records/marshmallow/literature/common/fulltext_links.py]:
```
"KEK scanned document"
```

File: backend/inspirehep/records/marshmallow/literature/ui.py [hunk 6]
```diff
@@ -97,6 +103,18 @@ def get_formatted_earliest_date(self, data):
             return missing
         return format_date(earliest_date)
 
+    def get_fulltext_links(self, data):
+        field_data = []
+        fields_to_include = {
+            "arxiv_eprints": ArxivFulltextLink(),
+            "external_system_identifiers": KekFulltextLink(),
+            "documents": InternalFulltextLink(),
+        }
```
Comment (drjova) [inline backend/inspirehep/records/marshmallow/literature/ui.py]: IMHO we shouldn't use schema in schema, why we just create three methods that handle this? i.e.

```python
def get_arxiv_fulltext_link():
   ....

def get_kek_fulltext_link():
   ....

def get_internal_fulltext_link():
   ....
```

Comment (ammirate) [reply backend/inspirehep/records/marshmallow/literature/ui.py]: @drjova what's the advantage of this?

Comment (drjova) [reply backend/inspirehep/records/marshmallow/literature/ui.py]: we're not calling ``dump`` schema in schema and it's just what we need

Comment (pazembrz) [reply backend/inspirehep/records/marshmallow/literature/ui.py]: IMHO your sollution will be even more complex. Here we have full info about the data source (so it if it arxiv, kek or internal url) so we do not kave to check almost anything. Inside Serializer we don't, so this will require more checks like: if keys are called `description` and `value` or are they called `schema` and `value`. Additionally recognition of schema KEK vs arXiv will require more logic inside this simple serializer. So there will be quire complex logic for simple thing.
Also Moving this amount of logic so deep into serializers is really bad idea.

Comment (drjova) [reply backend/inspirehep/records/marshmallow/literature/ui.py]: it's the wrong way to use marshmallow, if you want to put the logic in a ``schema`` you can use ``Nested``, calling ``schema`` in ``schema`` it's wrong. Which means ``pre_dump`` you analyzing the date and you let marshmallow to handle the dumping.

Comment (pazembrz) [reply backend/inspirehep/records/marshmallow/literature/ui.py]: But wait. You want to serialize it without serializer? Just pure dict in here? Then why we use serializes at all? Is this really what we want?

File: backend/tests/integration/records/indexer/test_indexer_literature.py [hunk 7]
```diff
@@ -72,3 +68,49 @@ def test_indexer_deletes_record_from_es(es_clear, db, datadir, create_record):
     hits_total = response["hits"]["total"]
 
     assert hits_total == expected_total
+
+
+@patch("inspirehep.records.api.literature.LiteratureRecord.add_files")
+def test_indexer_creates_proper_fulltext_links_in_ui_display(
+    add_file_mock, base_app, es_clear, db, create_record
```
Comment (drjova) [inline backend/tests/integration/records/indexer/test_indexer_literature.py]: We have ``enable_files`` AFAIR, we can use that :)

Comment (pazembrz) [reply backend/tests/integration/records/indexer/test_indexer_literature.py]: :+1:

File: backend/tests/integration/records/indexer/test_indexer_literature.py [hunk 8]
```diff
@@ -72,3 +68,49 @@ def test_indexer_deletes_record_from_es(es_clear, db, datadir, create_record):
     hits_total = response["hits"]["total"]
 
     assert hits_total == expected_total
+
+
+@patch("inspirehep.records.api.literature.LiteratureRecord.add_files")
+def test_indexer_creates_proper_fulltext_links_in_ui_display(
```
Comment (harunurhan) [inline backend/tests/integration/records/indexer/test_indexer_literature.py]: we might need a test case when all the source fields are empty, and we don't return `fulltext_links` at all.

also tests for the detail view would be good to

**Comment by harunurhan** [comment]
commit message is misleading, as this doesn't only `_ui_display`, but also detail pages. `literature: ` prefix would be more appropriate.

File: backend/tests/integration/records/indexer/test_indexer_literature.py (hunk 7 above)
Comment (pazembrz) [reply backend/tests/integration/records/indexer/test_indexer_literature.py]: It only sets flag in config and looks like `enable_files` is a little broken, it's not restoring previous value in config  when test ends.

File: backend/tests/integration/records/indexer/test_indexer_literature.py (hunk 8 above)
Comment (pazembrz) [reply backend/tests/integration/records/indexer/test_indexer_literature.py]: :+1:
